net processing: swap out signals for an interface class

theuni opened the pull request:
See individual commits.
Benefits:
- Allows us to begin moving stuff out of CNode and into CNodeState (after #10652 and follow-ups)
- Drops boost dependency and overhead
- Drops global signal registration
- Friendlier backtraces

theuni committed:
net: use an interface class rather than signals for message processing

Drop boost signals in favor of a stateful class. This will allow the message
processing loop to actually move to net_processing in a future step.

theuni committed:
net: pass CConnman via pointer rather than reference

There are a few too many edge-cases here to make this a scripted diff.

The following commits will move a few functions into PeerLogicValidation, where
the local connman instance can be used. This change prepares for that usage.

theuni committed:
net: drop unused connman param

The copy in PeerLogicValidation can be used instead.

gmaxwell commented:
Super mega concept ACK.  FWIW, I tested something functionally like this before (eliminating the signals with a direct function call) and saw a measurable performance improvement.

TheBlueMatt commented:
This is awesome. Lots of future work to do, but good first step. Will circle back around post-15.

jnewbery commented:
@gmaxwell 

> tested something functionally like this before (eliminating the signals with a direct function call) and saw a measurable performance improvement

How did you test that? It'd be good to benchmark whether this gives us a performance improvement.

jnewbery commented on src/net.cpp:
```diff
@@ -1996,15 +1996,16 @@ void CConnman::ThreadMessageHandler()
                 continue;
 
             // Receive messages
```
Nit: move `Receive messages` comment into the if block (the block covers both receiving and sending messages. Having the comment outside the block suggests it's just for receiving messages)

jnewbery commented on src/net.h:
```diff
@@ -422,18 +423,18 @@ struct CombinerAll
 };
 
 // Signals for message handling
```
Change comment to `Interface for message handling`

jnewbery commented on src/net_processing.h:
```diff
@@ -22,22 +22,32 @@ static const unsigned int DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN = 100;
+    * Send queued protocol messages to be sent to a give node.
+    *
+    * @param[in]   pto             The node which we are sending messages to.
+    * @param[in]   connman         The connection manager for that node.
```
remove this line. connman is no longer a param for this function.

jnewbery reviewed (commented):
Tested ACK f785e8479a8ffca6c89d71f9dd6f1c1b455c2671. A few nitty comments inline, plus one more: https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L126 seems to be a comment that's got stranded from its code, and is now nonsensical after this PR. Consider removing it entirely.

I've run this over the function test suite and it seemed to run slightly faster, though within expected variance. Do you have any ways of more focused performance testing?

theuni commented:
Remembering @gmaxwell's mention of it before, I was hoping to see a slight speedup here, but I haven't been able to observe anything substantial either. No test varies more than 1% from any other.

@jnewbery my test was to do a sync to block 150000 between two localhost nodes running from ramdisk datadirs. I believe that focuses on the hot-spot well enough.

theuni commented:
Needed rebase after #10179, so I went ahead and squashed it down too.

gmaxwell commented:
@jnewbery GBE connected hosts running a patch like this, out of ram, may have also required turning up our networking buffers. maybe that other locking changes we made mooted the improvement. I'll go see if I can find the relevant chat logs.

gmaxwell reviewed (approved):
ACK.

TheBlueMatt commented on src/net.cpp:
```diff
@@ -1995,16 +1995,17 @@ void CConnman::ThreadMessageHandler()
+                    return;
+                // Send messages
+                {
+                    LOCK(pnode->cs_sendProcessing);
```
Can you drop cs_sendProcessing while you're here? Its literally entirely unused except for this lock here.

theuni replied:
Sure. It's safe to drop it now, but we'd need to put it back for parallel processing.

TheBlueMatt commented on src/net.cpp:
```diff
@@ -2212,6 +2213,7 @@ CConnman::CConnman(uint64_t nSeed0In, uint64_t nSeed1In) : nSeed0(nSeed0In), nSe
     nMaxAddnode = 0;
     nBestHeight = 0;
     clientInterface = NULL;
+    messageInterface = NULL;
```
I believe we're nullptr'ing now?

theuni replied:
ok

TheBlueMatt commented on src/net.cpp:
```diff
@@ -1966,7 +1964,9 @@ bool CConnman::OpenNetworkConnection(const CAddress& addrConnect, bool fCountFai
 
-    GetNodeSignals().InitializeNode(pnode, *this);
+    if (messageInterface) {
+        messageInterface->InitializeNode(pnode);
```
I feel like we should just go for the direct deref here. A CConnman without a messageInterface feels like a client-of-the-API failure. Maybe just assert(messageInterface) in ::Start?

theuni replied:
Heh, I went back and forth on this. Was waiting for someone to complain. Agreed.

TheBlueMatt commented on src/net.h:
```diff
@@ -421,19 +422,19 @@ struct CombinerAll
-    boost::signals2::signal<void (CNode*, CConnman&)> InitializeNode;
-    boost::signals2::signal<void (NodeId, bool&)> FinalizeNode;
+protected:
+    NetEventsInterface() = default;
```
Why did you define the constructor here?

theuni replied:
Just a habit of mine to declare the ctor as protected for classes that should be inherited from. Since there are pure virtuals here, it's not really necessary.

TheBlueMatt commented on src/net.h:
```diff
@@ -421,19 +422,19 @@ struct CombinerAll
+protected:
+    NetEventsInterface() = default;
+public:
+    virtual ~NetEventsInterface() = default;
```
Why do we need to make the destructor virtual? CConnman doesn't care about construction/destruction, it just takes a ptr.

theuni replied:
To make sure that the base gets cleaned up if, for example, it's stored as an abstract std::unique_ptr<NetEventsInterface>

TheBlueMatt replied:
Well my point was the CConnman (and related net stuff), which "owns" this interface explicitly does not do so (and does not have any desire to ever own the object), so adding anything additional to the interface feels strange to me.

theuni replied:
I must be missing your point. This is just a safety precaution in case anything ever wants to hold an abstract pointer. This may be the case if, for example, we create a vector of message processors in order to run them against each-other.

I can remove it if you'd like, but I'm not understanding the objection to being cautious.

TheBlueMatt replied:
Well my objection was just that it forces you into a virtual destructor unless you remember to call derived classes final, and since I dont think anything should ever have anything other than a reference to (ie will not destruct directly) a NetEventsInterface, no need to make anything virtual. I don't care too strongly, however.

TheBlueMatt reviewed (commented):
This is awesome. As mentioned, after this, PeerLogicValidation should probably get a raname, as the "Validation" referred to ValidationInterface.

TheBlueMatt commented:
Note that some of the Process functions are likely to move again into CNodeState or so, I believe, but this is an obviously good step.

theuni commented:
Rebased, updated for recent changes, and addressed @TheBlueMatt's feedback

TheBlueMatt reviewed (commented):
utACK 6e6f06c8d971f21262c6f2272b70170d19670c2a. Feel like renaming PeerLogicValidation now that its not strictly about "Validation" anymore?

theuni commented:
Rebased. Same as before, plus I took the suggestion here: https://github.com/bitcoin/bitcoin/pull/10756#discussion_r133860799

laanwj commented on src/net.cpp:
```diff
@@ -89,10 +89,6 @@ std::string strSubVersion;
 
 limitedmap<uint256, int64_t> mapAlreadyAskedFor(MAX_INV_SZ);
 
-// Signals for message handling
```
Another global block gone, nice!

laanwj commented on src/net.h:
```diff
@@ -156,11 +156,13 @@ class CConnman
         nMaxOutboundTimeframe = connOptions.nMaxOutboundTimeframe;
         nMaxOutboundLimit = connOptions.nMaxOutboundLimit;
         vWhitelistedRange = connOptions.vWhitelistedRange;
+
```
Empty line?

theuni replied:
Fixed.

laanwj commented on src/net.h:
```diff
@@ -436,19 +439,16 @@ struct CombinerAll
 
-// Signals for message handling
-struct CNodeSignals
+// Interface for message handling
```
Ideally we'd use a doxygen-compliant comment here, so that it shows the class description
```
/**
 * Interface for message handling
 */
```

theuni replied:
I had to rebase anyway, so I went ahead and fixed these two nits as well.

JeremyRubin commented on src/net_processing.h:
```diff
@@ -22,22 +22,31 @@ static const unsigned int DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN = 100;
-void UnregisterNodeSignals(CNodeSignals& nodeSignals);
-
-class PeerLogicValidation : public CValidationInterface {
+class PeerLogicValidation : public CValidationInterface, public NetEventsInterface {
```
Maybe it makes sense to mark stuff final/not be overriding an abc here? NetEventsInterface feels overkill  unless we anticipate per-connection different handlers.

theuni replied:
I don't anticipate per-connection, but definitely per-connman. I have immediate plans to add different handlers and run them against each-other for testing.

Making this final is a good idea, but I'd rather not keep squashing more changes in. Let's do that in a follow up that also renames PeerLogicValidation to something that better reflects what it's doing.

JeremyRubin reviewed (approved):
utack 80e2e9d

TheBlueMatt commented on src/init.cpp:
```diff
@@ -195,11 +195,10 @@ void Shutdown()
     UnregisterValidationInterface(peerLogic.get());
-    peerLogic.reset();
     g_connman.reset();
+    peerLogic.reset();
```
As discussed, peerLogic should probably get destroyed before g_connman, though g_connman needs to be stopped first (and all references to m_msgproc in CConnman need to be in threads that will be stopped or should be gated on running - OpenNetworkConnection I believe is the only current violation of this).

theuni committed:
net: stop both net/net_processing before destroying them

This should avoid either attempting to use an invalid reference/pointer to the
other.

TheBlueMatt commented:
utACK 2525b972af6645ca239ac1078cffb132b402bfbb
